ad-joining/register-computer/ad/domain.py

```diff
@@ -79,23 +80,37 @@ def locate_domain_controllers(domain_name, site_name):
         return [str(record.target)[:-1] if str(record.target).endswith(".") else str(record.target) for record in records_sorted]
 
     @staticmethod
-    def connect(domain_controller, base_dn, user, password):
-        logging.info("Connecting to LDAP endpoint of '%s' as '%s'" % (domain_controller, user))
-        connection = ldap3.Connection(
-            server=ldap3.Server(
-                domain_controller,
-                port=389,
-                connect_timeout=5),
-            user=user,
-            password=password,
-            authentication=ldap3.NTLM,
-            raise_exceptions=True)
-
-        if not connection.bind():
-            raise LdapException("Connecting to LDAP endpoint of %s as '%s' failed, check credentials" %
-                (domain_controller, user))
-
-        return ActiveDirectoryConnection(domain_controller, connection, base_dn)
+    def connect(domain_controller, base_dn, user, password, certificate_data=None):
+        logging.info("Connecting to LDAPS endpoint of '%s' as '%s'" % (domain_controller, user))
+        tls_configuration = Tls(ssl.create_default_context(ssl.Purpose.SERVER_AUTH), validate=ssl.CERT_REQUIRED)
+
+        if not certificate_data is None:
+            logging.debug("Using CA certificate data from Secret Manager")
+            tls_configuration.ca_certs_data = certificate_data
+
+        server = ldap3.Server(domain_controller, port=636, connect_timeout=5, use_ssl=True, tls=tls_configuration)
```

Comment:
It's a bit expensive to try LDAP/S first every time. If there is a firewall in between, it's possible that connection attempts don't fail immediately, but cause a 5sec timeout.

Can we use the config to determine whether to use LDAP or LDAP/S? I guess we'd need another setting for that...

Reply:
I added a USE_LDAPS flag to decide whether using LDAPS or LDAP.